### PR TITLE
Allow text buffer size, shadow offset and radius to be set in "percentage" units

### PR DESCRIPTION
**Pull request by nyalldawson**
Allows these sizes to be set as a percent of the font size. This is
desirable as it allows creation of text formats where the components
nicely scale as the text size is changed, instead of having
constant buffer/shadow sizes which need to be independantly adjusted
when the text size changes.

### CODE REVIEW

File: src/core/textrenderer/qgstextformat.cpp [hunk 1]
```diff
@@ -1095,15 +1095,16 @@ QPixmap QgsTextFormat::textFormatPreviewPixmap( const QgsTextFormat &format, QSi
   context.setFlag( Qgis::RenderContextFlag::Antialiasing, true );
 
   // slightly inset text to account for buffer/background
+  const double fontSize = context.convertToPainterUnits( tempFormat.size(), tempFormat.sizeUnit(), tempFormat.sizeMapUnitScale() );
   double xtrans = 0;
   if ( tempFormat.buffer().enabled() )
-    xtrans = context.convertToPainterUnits( tempFormat.buffer().size(), tempFormat.buffer().sizeUnit(), tempFormat.buffer().sizeMapUnitScale() );
+    xtrans = tempFormat.buffer().sizeUnit() == QgsUnitTypes::RenderPercentage ? fontSize * tempFormat.buffer().size() / 100 : context.convertToPainterUnits( tempFormat.buffer().size(), tempFormat.buffer().sizeUnit(), tempFormat.buffer().sizeMapUnitScale() );
```
Comment (nirvn) [inline src/core/textrenderer/qgstextformat.cpp:1101]: It's a fairly long line, could you break it up the way you did in qgstextrenderer.cpp at line 308?

File: src/core/textrenderer/qgstextformat.cpp [hunk 2]
```diff
@@ -1095,15 +1095,16 @@ QPixmap QgsTextFormat::textFormatPreviewPixmap( const QgsTextFormat &format, QSi
...
   if ( tempFormat.background().enabled() && tempFormat.background().sizeType() != QgsTextBackgroundSettings::SizeFixed )
     xtrans = std::max( xtrans, context.convertToPainterUnits( tempFormat.background().size().width(), tempFormat.background().sizeUnit(), tempFormat.background().sizeMapUnitScale() ) );
 
   double ytrans = 0.0;
   if ( tempFormat.buffer().enabled() )
-    ytrans = std::max( ytrans, context.convertToPainterUnits( tempFormat.buffer().size(), tempFormat.buffer().sizeUnit(), tempFormat.buffer().sizeMapUnitScale() ) );
+    ytrans = std::max( ytrans, tempFormat.buffer().sizeUnit() == QgsUnitTypes::RenderPercentage ? fontSize * tempFormat.buffer().size() / 100 : context.convertToPainterUnits( tempFormat.buffer().size(), tempFormat.buffer().sizeUnit(), tempFormat.buffer().sizeMapUnitScale() ) );
```
Comment (nirvn) [inline src/core/textrenderer/qgstextformat.cpp:1107]: Same

File: tests/src/python/test_qgstextrenderer.py [hunk 3]
```diff
@@ -2285,6 +2285,16 @@ def testDrawBufferSizePixels(self):
         format.buffer().setSizeUnit(QgsUnitTypes.RenderPixels)
         assert self.checkRender(format, 'text_buffer_pixels', QgsTextRenderer.Buffer, text=['test'])
 
+    def testDrawBufferSizePercentage(self):
```
Comment (nirvn) [inline tests/src/python/test_qgstextrenderer.py:2288]: Might be worth testing this % unit type against a data-defined font size property if the checkRender can handle this.

**Review by nirvn — APPROVED** [review]
Nice one.

File: tests/src/python/test_qgstextrenderer.py (hunk 3 above)
Comment (nyalldawson) [reply tests/src/python/test_qgstextrenderer.py:2302]: Data defined properties with text format doesn't really work that way -- you call updateDataDefinedProperties() on the text format itself and all its properties are updated in place

Comment (nirvn) [reply tests/src/python/test_qgstextrenderer.py:2302]: @nyalldawson , roger that.

**Pull request merged**